server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -757,6 +786,84 @@ public DataStream addBackingIndex(Metadata clusterMetadata, Index index) {
         );
     }
 
+    /**
+     * Adds the specified index as a failure store index and returns a new {@code DataStream} instance with the new combination
+     * of failure store indices.
+     *
+     * @param index index to add to the data stream's failure store
+     * @return new {@code DataStream} instance with the added failure store index
+     * @throws IllegalArgumentException if {@code index} is ineligible to be a failure store index for the data stream
+     */
+    public DataStream addFailureStoreIndex(Metadata clusterMetadata, Index index) {
+        // validate that index is not part of another data stream
+        final var parentDataStream = clusterMetadata.getIndicesLookup().get(index.getName()).getParentDataStream();
+        if (parentDataStream != null) {
+            validateDataStreamAlreadyContainsIndex(index, parentDataStream, true);
+            return this;
+        }
+
+        ensureNoAliasesOnIndex(clusterMetadata, index);
+
+        List<Index> updatedFailureIndices = new ArrayList<>(failureIndices);
+        updatedFailureIndices.add(0, index);
+        assert updatedFailureIndices.size() == failureIndices.size() + 1;
+        return new DataStream(
+            name,
+            indices,
+            generation + 1,
+            metadata,
+            hidden,
+            replicated,
+            system,
+            allowCustomRouting,
+            indexMode,
+            lifecycle,
+            failureStore,
+            updatedFailureIndices,
+            autoShardingEvent
```

Comment:
Depending which PR gets merged first, we should ensure we copy here the `rolloverOnWrite` too :)

Reply:
I've just marked my PR for `auto-merge`, so it should go through anytime soon.